cumulocity-blocks/CreateMultiMeasurement.mon

```diff
@@ -0,0 +1,153 @@
+ */
+
+/*********
+ * Note: this block is now available in 10.9.0.x or later of Analytics Builder as
+ * part of the product.  The below is for reference use only, we recommend using
+ * the product block where applicable.
+ ********/
+
+package apamax.analyticskit.blocks.cumulocity;
+
+using apama.analyticsbuilder.BlockBase;
+using apama.analyticsbuilder.Activation;
+using apama.analyticsbuilder.Partition_Broadcast;
+
+using com.apama.cumulocity.Measurement;
+using com.apama.cumulocity.MeasurementValue;
+using com.apama.exceptions.Exception;
+using apama.analyticsbuilder.cumulocity.CumulocityOutputParams;
+using apama.analyticsbuilder.cumulocity.CumulocityOutputHandler;
+using apama.analyticsbuilder.Promise;
+using apama.analyticsbuilder.Value;
+
+event CreateMultiMeasurement_$Parameters {
+	/**
+	 * Device or Trigger Device.
+	 *
+	 * The device (or for models handling groups, Trigger Device) to which the measurement is to be sent.
+	 *
+	 * The model editor uses the current device name. This is mapped internally to the device identifier.
+	 * @$semanticType c8y_deviceIdOrCurrentDevice
+	 */
+	any deviceId;
+
+	/**
+	 * Measurement Type.
+	 *
+	 * The type of the measurement.
+	 */
+	string measurementType;
+
+	/** Validate parameters */
+	action $validate() {
+		switch(deviceId) {
+			case string: { BlockBase.throwsOnEmpty(deviceId, "deviceId", self); }
+			default: {
+				if deviceId = any() or deviceId.getEntry("currentDevice") != <any> true {
+					throw Exception("Expected a device Id or a currentDevice object", "IllegalArgumentException"); // NON-L10N-DEV
+				}
+			}
+		}
+		BlockBase.throwsOnEmpty(measurementType, "measurementType", self);
+	}
+}
+
+/**
+ * Complex Measurement Output.
+ *
+ * Creates a new complex <tt>Measurement</tt> object for the associated device or the triggering device.
+ *
+ * This block sends a new measurement to Cumulocity IoT containing multiple fragments and series. 
+ * The list of fragments and series is included as properties on the Value input. The measurement
+ * is sent to the current device or the device specified.
+ *
+ * Non-finite values are ignored.
+ *
+ * @$blockCategory Output
+ * @$producesOutput
+ * @$blockType c8y_Output
+ * @$derivedName $measurementType Measurement
+ * @$titleIsDerived true
+ */
+event CreateMultiMeasurement {
+	BlockBase $base;
+	CreateMultiMeasurement_$Parameters $parameters;
+	/** Sends to the 'current' device from the activation's partition: */
+	CumulocityOutputHandler outputHandler;
+	
+	action $validate() returns Promise {
+		dictionary<string, any> fields := {"type":<any>$parameters.measurementType};
+	    // Creating an CumulocityOutputParams object
+	    CumulocityOutputParams params := CumulocityOutputParams.forSyncEventType($parameters.deviceId, 
+	                                            self, Measurement.getName(), fields);
```

Comment:
Similar approach to alarms and events to use the type and not fragment and series like for the normal measurement input block. Ok?

Reply:
The framework won't be able to figure out chains between models. It uses same sets of properties for an event type to figure out if outputs from an output block are consumed by any input blocks. For measurements, the product blocks use source, fragment and series to figure this out.

So, fine only if you do not want any chains between models.

Reply:
Is there an alternative? A measurement produced by the block will have many series. I did not see how I could model that here.

Reply:
If all fragment+series to output are known beforehand then you can declare them all using multiple `CumulocityOutputParams`. In your block it seems to be dynamic so I don't think we can do much. Or you can allow dynamic number of fragment+series to be configured via parameters using something like `sequence<NameValue>`, similar to the range lookup - https://github.com/SoftwareAG/apama-analytics-builder-block-sdk/blob/main/doc/040-Parameters.md#parameters-block-startup-and-error-handling

Maybe analytics builder runtime should be enhanced as well so that it can form chain using subset relation as well in addition to exact equality. For example, if model1 declares that it produces Measurements for device D1 without specifying anything about the fragment and series, and model2 declares that it consumes Measurements for device D1 for fragment+series F.S, the chain is still formed between model1 and model2 because model1 can produce measurements which can be consumed by model2.
Other fix would be to enhance the framework so that it does